File: src/levels/L04-singleton/strategy.rb
class TicketCounter
  def initialize = @last = 0

  # region: TicketCounter#next
  def next = @last += 1
  # endregion
end

class Register
  # La caja elige una estrategia de numeración… y cada estrategia lleva su propia cuenta.
  def initialize(display, numbering = TicketCounter.new)
    @display = display
    @numbering = numbering
  end

  # region: Register#checkout
  def checkout(order)
    order.ticket = @numbering.next
    @display.show(order)
  end
  # endregion
end

File: src/levels/L04-singleton/ts/strategy.ts
interface Numbering {
  next(): number
}

class TicketCounter implements Numbering {
  private last = 0

  // region: TicketCounter#next
  next() { return ++this.last }
  // endregion
}

class Register {
  // La caja elige una estrategia de numeración… y cada estrategia lleva su propia cuenta.
  constructor(
    private display: Display,
    private numbering: Numbering = new TicketCounter(),
  ) {}

  // region: Register#checkout
  checkout(order: Order) {
    order.ticket = this.numbering.next()
    this.display.show(order)
  }
  // endregion
}
